### PR TITLE
[DataLayout] Refactor the rest of `parseSpecification`

### PR DESCRIPTION
**Pull request by s-barannikov**
The aim is to improve test coverage of data layout string parsing.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/DataLayoutImporter.cpp [hunk 1]
```diff
@@ -197,9 +197,9 @@ DataLayoutImporter::tryToEmplaceStackAlignmentEntry(StringRef token) {
   if (failed(alignment))
     return failure();
 
-  // Only store the stack alignment if it has a non-default value.
+  // Stack alignment shouldn't be zero.
   if (*alignment == 0)
-    return success();
+    return failure();
```
Comment (s-barannikov) [inline mlir/lib/Target/LLVMIR/DataLayoutImporter.cpp:202]: @gysit 
Please take a look if this change is ok.
Previously, "S0" in data layout string was equivalent to omitting it. After this patch, specifying zero for alignment is not allowed.

Comment (gysit) [reply mlir/lib/Target/LLVMIR/DataLayoutImporter.cpp:202]: Looks good to me!

**Comment by s-barannikov** [comment]
@nikic Please take a look when you have time. This is the last part.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**